### PR TITLE
add torch_dtype control for `prepare_model_for_kbit_training`

### PR DESCRIPTION
**Pull request by ranchlai**
# What does this PR do

This pr add torch_dtype parameter to `prepare_model_for_kbit_training` for better control of non-quantized modules. 

### Why I think this is necessary
The [bnb lib](https://github.com/ranchlai/bitsandbytes/blob/main/bitsandbytes/autograd/_functions.py#L322C31-L322C31) `MatMul8bitLt` will use `float16` for internal computing. It should be better if we keep the non-int8 in float16, especially when the model is already in half precision. 

`torch_dtype` is set to `torch.float32` by default. Should be ok for existing code. 

Please leave comments. ^^

### CODE REVIEW

**Comment by ranchlai** [comment]
Still running some test with deepspeed.

**Comment by BenjaminBossan** [comment]
Thanks for the addition. I agree it's a good idea to give users more control about the dtype in this situation. When you're finished testing, let us know.

File: src/peft/utils/other.py [hunk 1]
```diff
@@ -95,10 +95,10 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
         param.requires_grad = False
 
     if not is_gptq_quantized:
-        # cast all non INT8 parameters to fp32
+        # cast all non INT8 parameters to torch_dtype
         for param in model.parameters():
-            if (param.dtype == torch.float16) or (param.dtype == torch.bfloat16):
-                param.data = param.data.to(torch.float32)
+            if param.dtype not in [torch.int8, torch.uint8]:
```
Comment (pacman100) [inline src/peft/utils/other.py:100]: Does this check work with 4-bit/3-bit quantization?

**Review by pacman100 — COMMENTED** [review]
Hello @ranchlai, Thank you for the PR. The main reason for casting non-quantized weights to FP32 is given in the PR https://github.com/huggingface/peft/pull/389 which is copied below:

a. When making modules trainable via `modules_to_save`, it was resulting in loss being `NaN ` or `0` because those params are in fp16/bf16 which are unstable for training.
b. half and float mismatches because certain layernorm were being converted into FP32 and when making other trainable modules to be in FP32 manually, it was leading to mismatches as frozen layers were in half-precision and few in full precision.

The addition of `dtype` param has to be used with caution by end-user and as such I feel this might lead to more issues and confusion. For stability, LayerNorm layers should be in FP32, but with the `dtype` param they would end up in half-precision which again might lead to unexpected results.

File: src/peft/utils/other.py (hunk 1 above)
Comment (ranchlai) [reply src/peft/utils/other.py:100]: @pacman100. Thanks for the info. Indeed, setting to fp16  will lead to  Nan and loss 0, and there is not any warning or error. 

I spent a few hours trying to find out why. I am using llama2 for lora finetuning, using peft+accelerate(+deepspeed) here is what I found out, for now:
- overflow is found in the first forward pass of my training process,  not even reaching the backward pass
-  bnb is OK doing its job.
- RMSNorm for llama is OK, as it will  cast input to float32 and back.
- attention is OK , 4096 x 4096 seems ok for my test
- Overflow found in MLP, due to the up projection matrix multiplication, 11008x4096 matrix.

Comment (ranchlai) [reply src/peft/utils/other.py:100]: 4-bit is stored as torch.uint8 as I know.  but I a have no idea about 3bit. ?

**Comment by BenjaminBossan** [comment]
In general, I would be in favor of giving the user more control but as @pacman100 explained well, casting to half precision could lead to several issues down the line.

I wonder if an alternative could be that we just add an option to skip casting to float32 entirely. We would add to the docs that users should choose that option only if they know what they're doing and want to cast the layers manually. Then it's their responsibility to choose the right dtype for the right layers.

**Comment by ranchlai** [comment]
For fp16,  I can make it work but need to clamp the MLP layer.  For bf16, it should be safe. Still waiting for the training to finish
@pacman100 I am thinking why the training is unstable under fp16. There are many efforts to make training stable such as gradient clipping and loss scaling. Maybe we should also add hidden-state clipping in transformer or peft lib?

**Comment by ranchlai** [comment]
> 
@BenjaminBossan Thanks very much for comments.  I think that's ok for me, but its up to you guys. 

PS: I think allowing model to be in lower precision is a trend, even in training.

**Comment by ranchlai** [comment]
I think I will close this for now, maybe re-open if my bf16/fp16 training runs well.  Thanks all for replying.
